### PR TITLE
nix.tests.compareEval

### PR DESCRIPTION
**Pull request by roberth**
## Description of changes

Instantiate all of Nixpkgs when Nix is updated.
Succeed only when the resulting derivations are the same as Nix 2.13.

This test will also be used in upstream CI.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Ma27** [comment]
Without having reviewed it yet, I want to say that I very much appreciate this change!

File: pkgs/tools/package-management/nix/tests/compareEval.nix
Comment (RaitoBezarius) [inline pkgs/tools/package-management/nix/tests/compareEval.nix:25]: Why this choice of reference Nix? As we do have minVer = 2.3, shouldn't we use Nix 2.3?

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
Thank you for the awesome code, shouldn't the evaluation machinery be generalized out of the test directory and be shared with the evaluation entrypoints for all of nixpkgs, cc @lilyinstarlight with whom we had discussions regarding reforming the evaluation entrypoints of nixpkgs.

Just to be sure, this idea of performing sharding has first surfaced in an Adam Joseph's PR, I think it's good to credit them somewhere in the file, pointing to their defunct PR or to point to the related work regarding this.

His (brilliant) idea was to accelerate further the evaluation of nixpkgs using sharding and GNU parallel, which was an implementation of nix-eval-jobs in pure Nix + 1 Bash utility.

File: pkgs/tools/package-management/nix/tests/compareEval.nix
Comment (roberth) [reply pkgs/tools/package-management/nix/tests/compareEval.nix:25]: Nix 2.3 had different logic for printing floats, and it's the only currently packaged Nix that has the old behavior.
By making 2.13 the reference version, we can check all attributes without making any exceptions (`knownIssues`).
Updates of 2.3 will be tested against 2.13, and vice versa (although that is highly unlikely to happen, because both are EOL and 2.13 doesn't receive the attention that 2.3 still does).

**Comment by roberth** [comment]
The only goal for this PR is to test Nix; not Nixpkgs. The whole of Nixpkgs mostly only serves as test data here.

> shouldn't the evaluation machinery be generalized out of the test directory and be shared with the evaluation entrypoints for all of nixpkgs

I don't know. It would be nice to have an recognizable interface into this logic, so that Nix upstream can reuse it without it being broken every now and then.
On the other hand, this stuff is rather ad hoc and not something I think people want to use, unless they're developing or maintaining Nix, or an alternate implementing like tvix, I suppose, so maybe it's not worth the effort of formalizing everything, and then having more inertia later. It's easy enough to adapt when something does change.


> be shared with the evaluation entrypoints for all of nixpkgs, cc @lilyinstarlight with whom we had discussions

I'm not always up to date with the Nixpkgs Architecture Team room, but did you share this project there? You could probably get some help that way.



> Just to be sure, this idea of performing sharding has first surfaced in an Adam Joseph's PR

I didn't follow that project much, and I reusing that work didn't seem like a good idea because I'd have to communicate with a banned contributor, and I would rather familiarize myself with Nix's intricacies than his implementation's intricacies, because the prior provides a more valuable experience as a Nix maintainer, but to be honest I'm not sure if I even put that much thought into the decision to only look at merged work to build on. Maybe I was just dumb. I did this work 1.5 months ago, so please forgive me for having to reverse engineer my thought process. I only continued it this week because my nix-diff PR got merged. Diagnostics are pretty important, and I have a lot on my plate, so ended up waiting for that.


> think it's good to credit them somewhere in the file

I haven't used his unmerged work, and we don't normally credit the authors of the code we reuse. I guess I could mention it in a commit message, as prior art or something, but I don't even know what to say, except something "Thanks to Adam Joseph for figuring out all-attrs-eval-under-tryEval, which was a great starting point for producing a list of Nixpkgs attributes".



> His (brilliant) idea was to accelerate further the evaluation of nixpkgs using sharding and GNU parallel, which was an implementation of nix-eval-jobs in pure Nix + 1 Bash utility.

GNU parallel is a pretty obvious choice for a shardable workload. Maybe it's brilliant; I haven't seen it.

Anyway, I think using separate derivations is sufficient. For what it's worth, it gives us distributed evaluation through remote builds; not just local parallelism.

File: pkgs/test/eval/all-attributes-eval-shard.nix
Comment (RaitoBezarius) [inline pkgs/test/eval/all-attributes-eval-shard.nix:7]: This should be a parameter, not everything evals the same on every platform.

Comment (RaitoBezarius) [reply pkgs/test/eval/all-attributes-eval-shard.nix:7]: And it should use localSystem and crossSystem.

**Comment by roberth** [comment]
> > I'm not always up to date with the Nixpkgs Architecture Team room, but did you share this project there? You could probably get some help that way.
> 
> And the fact that even me feels a bit offended about your response will definitely not give anyone the desire to venture in that room, it is a sad thing w.r.t. new contributors.

I was just asking and trying to be helpful. I suppose this could be read in a condescending tone. That's not at all what I was going for.
Talking to more people who care about such issues just seemed like good idea.




> I don't know where did you get the information they were a banned contributor, but they are not.

Oh my, oh my, was I wrong. He took a break on his own accord, and wasn't banned at all.
I don't know how I managed to misremember this... I'll correct that right away.

**Comment by roberth** [comment]
> I am a bit disturbed by the methodology you are hinting at

My methodology was just "solve one problem", open a PR and take it from there.
Whether that's simple or simplistic is your call, but know that I have no intent, and I do care about the big picture.

I am very sorry that I made a mistake earlier by ignoring his work, and I apologize for making you spend a whole lot of energy that should not have been necessary.

I'll take a few steps back and re-do this, or parts of this, or something similar, properly, in a new PR.

Thank you honestly very very much for your review; otherwise I would have remained oblivious to this serious mistake.

**Pull request closed without merging**
